From 7a9d3ed152cd7f0db82be95146b4d4baeb9a1e59 Mon Sep 17 00:00:00 2001
From: Allen Wild <allenwild93@gmail.com>
Date: Sat, 20 Jan 2018 01:44:55 -0500
Subject: [PATCH] ncurses: make ABI version user-selectable

The ncurses6 features can be useful and backwards-compatibility
with the ncurses5 ABI isn't always needed.

Also remove the narrowc libtinfo libraries so that widec versions get
installed to prevent runtime segfaults. Not sure why this is only needed
for ABI v6.

Signed-off-by: Allen Wild <allenwild93@gmail.com>
---
 meta/recipes-core/ncurses/ncurses.inc    | 14 +++++++++++---
 meta/recipes-core/ncurses/ncurses_6.4.bb |  2 +-
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/meta/recipes-core/ncurses/ncurses.inc b/meta/recipes-core/ncurses/ncurses.inc
index 3b72f3efdd..b707373cf9 100644
--- a/meta/recipes-core/ncurses/ncurses.inc
+++ b/meta/recipes-core/ncurses/ncurses.inc
@@ -22,6 +22,10 @@ CACHED_CONFIGUREVARS:append:linux = " cf_cv_working_poll=yes"
 
 EXTRASITECONFIG = "CFLAGS='${CFLAGS} -I${SYSROOT_DESTDIR}${includedir}'"
 
+# ncurses ABI version: 5 for compatibility with most external
+# applications, 6 for new features such as extra mouse buttons
+NCURSES_ABI_VERSION ?= "5"
+
 # Whether to enable separate widec libraries; must be 'true' or 'false'
 #
 # TODO: remove this variable when widec is supported in every setup?
@@ -174,8 +178,12 @@ do_install() {
                 install.data
 
 
-        ! ${ENABLE_WIDEC} || \
+        if ${ENABLE_WIDEC}; then
+            # remove libtinfo files to force the widec versions to get installed, otherwise
+            # segfaults happen at runtime
+            rm ${D}${libdir}/libtinfo*
             oe_runmake -C widec ${_install_cfgs} ${_install_opts}
+        fi
 
         cd narrowc
 
@@ -214,7 +222,7 @@ do_install() {
                 test -h $f || continue
                 rm -f $f
                 echo '/* GNU ld script */'  >$f
-                echo "INPUT($i.so.5 AS_NEEDED(-ltinfo))" >>$f
+                echo "INPUT($i.so.${NCURSES_ABI_VERSION} AS_NEEDED(-ltinfo))" >>$f
         done
 
         # Make sure that libcurses is linked so that it gets -ltinfo
@@ -239,7 +247,7 @@ do_install() {
 
             # Use ln -rs to ensure this is a relative link despite absolute paths
             # (as we can't know the relationship between base_libdir and libdir).
-            ln -rs ${D}${base_libdir}/libtinfo.so.5 ${D}${libdir}/libtinfo.so
+            ln -rs ${D}${base_libdir}/libtinfo.so.${NCURSES_ABI_VERSION} ${D}${libdir}/libtinfo.so
         fi
         if [ -d "${D}${includedir}/ncurses" ]; then
             for f in `find ${D}${includedir}/ncurses -name "*.h"`
diff --git a/meta/recipes-core/ncurses/ncurses_6.4.bb b/meta/recipes-core/ncurses/ncurses_6.4.bb
index 61558ecfa8..45a35258d5 100644
--- a/meta/recipes-core/ncurses/ncurses_6.4.bb
+++ b/meta/recipes-core/ncurses/ncurses_6.4.bb
@@ -12,7 +12,7 @@ SRC_URI += "file://0001-tic-hang.patch \
 # commit id corresponds to the revision in package version
 SRCREV = "1003914e200fd622a27237abca155ce6bf2e6030"
 S = "${WORKDIR}/git"
-EXTRA_OECONF += "--with-abi-version=5"
+EXTRA_OECONF += "--with-abi-version=${NCURSES_ABI_VERSION}"
 UPSTREAM_CHECK_GITTAGREGEX = "v(?P<pver>\d+_\d+)$"
 
 # This is needed when using patchlevel versions like 6.1+20181013
-- 
2.46.1

